## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.19.5'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [6]:
df = pd.concat([s1,s2], axis=1)
df.columns = ['col1','col2']
df

col1      col2
0   0.311711  0.324084
1   0.520068 -0.385082
2   0.546710 -0.676922
3   0.184854  0.611676
4   0.969585  1.031000
5   0.775133  0.931280
6   0.939499 -0.839218
7   0.894827 -0.309212
8   0.597900  0.331263
9   0.921874  0.975545
10  0.088493 -0.479174
11  0.195983 -0.185659
12  0.045227 -1.106335
13  0.325330 -1.196207
14  0.388677  0.812526
15  0.271349  1.356240
16  0.828738 -0.072010
17  0.356753  1.003533
18  0.280935  0.361636
19  0.542696 -0.645120

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [8]:
df[(df['col2'] >= 0) & (df['col2']<=1.0)]

col1      col2
0   0.311711  0.324084
3   0.184854  0.611676
5   0.775133  0.931280
8   0.597900  0.331263
9   0.921874  0.975545
14  0.388677  0.812526
18  0.280935  0.361636

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [13]:
df['col3'] = df['col2'].map(lambda x: 1 if x>=0 else -1)
df

col1      col2  col3
0   0.311711  0.324084     1
1   0.520068 -0.385082    -1
2   0.546710 -0.676922    -1
3   0.184854  0.611676     1
4   0.969585  1.031000     1
5   0.775133  0.931280     1
6   0.939499 -0.839218    -1
7   0.894827 -0.309212    -1
8   0.597900  0.331263     1
9   0.921874  0.975545     1
10  0.088493 -0.479174    -1
11  0.195983 -0.185659    -1
12  0.045227 -1.106335    -1
13  0.325330 -1.196207    -1
14  0.388677  0.812526     1
15  0.271349  1.356240     1
16  0.828738 -0.072010    -1
17  0.356753  1.003533     1
18  0.280935  0.361636     1
19  0.542696 -0.645120    -1

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [16]:
df['col4'] = df['col2'].clip(-1.0,1.0)
df

col1      col2  col3      col4
0   0.311711  0.324084     1  0.324084
1   0.520068 -0.385082    -1 -0.385082
2   0.546710 -0.676922    -1 -0.676922
3   0.184854  0.611676     1  0.611676
4   0.969585  1.031000     1  1.000000
5   0.775133  0.931280     1  0.931280
6   0.939499 -0.839218    -1 -0.839218
7   0.894827 -0.309212    -1 -0.309212
8   0.597900  0.331263     1  0.331263
9   0.921874  0.975545     1  0.975545
10  0.088493 -0.479174    -1 -0.479174
11  0.195983 -0.185659    -1 -0.185659
12  0.045227 -1.106335    -1 -1.000000
13  0.325330 -1.196207    -1 -1.000000
14  0.388677  0.812526     1  0.812526
15  0.271349  1.356240     1  1.000000
16  0.828738 -0.072010    -1 -0.072010
17  0.356753  1.003533     1  1.000000
18  0.280935  0.361636     1  0.361636
19  0.542696 -0.645120    -1 -0.645120

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [19]:
df['col2'].nlargest(5)

15    1.356240
4     1.031000
17    1.003533
9     0.975545
5     0.931280
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [20]:
df['col2'].nsmallest(5)


13   -1.196207
12   -1.106335
6    -0.839218
2    -0.676922
19   -0.645120
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [21]:
df.cumsum()

col1      col2  col3      col4
0   0.311711  0.324084     1  0.324084
1   0.831779 -0.060998     0 -0.060998
2   1.378489 -0.737920    -1 -0.737920
3   1.563344 -0.126244     0 -0.126244
4   2.532928  0.904756     1  0.873756
5   3.308061  1.836036     2  1.805036
6   4.247560  0.996818     1  0.965819
7   5.142388  0.687606     0  0.656606
8   5.740288  1.018869     1  0.987870
9   6.662162  1.994414     2  1.963415
10  6.750654  1.515240     1  1.484241
11  6.946637  1.329581     0  1.298582
12  6.991864  0.223246    -1  0.298582
13  7.317195 -0.972961    -2 -0.701418
14  7.705872 -0.160435    -1  0.111107
15  7.977221  1.195805     0  1.111107
16  8.805959  1.123795    -1  1.039097
17  9.162712  2.127328     0  2.039097
18  9.443646  2.488964     1  2.400733
19  9.986343  1.843844     0  1.755614

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [22]:
df['col2'].median()

0.1260369239072306

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [23]:
df[df['col2']>0.0]

col1      col2  col3      col4
0   0.311711  0.324084     1  0.324084
3   0.184854  0.611676     1  0.611676
4   0.969585  1.031000     1  1.000000
5   0.775133  0.931280     1  0.931280
8   0.597900  0.331263     1  0.331263
9   0.921874  0.975545     1  0.975545
14  0.388677  0.812526     1  0.812526
15  0.271349  1.356240     1  1.000000
17  0.356753  1.003533     1  1.000000
18  0.280935  0.361636     1  0.361636

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [24]:
df[:5].to_dict()

{'col1': {0: 0.31171107608941095,
  1: 0.5200680211778108,
  2: 0.5467102793432796,
  3: 0.18485445552552704,
  4: 0.9695846277645586},
 'col2': {0: 0.324083969394795,
  1: -0.38508228041631654,
  2: -0.6769220003059587,
  3: 0.6116762888408679,
  4: 1.030999522495951},
 'col3': {0: 1, 1: -1, 2: -1, 3: 1, 4: 1},
 'col4': {0: 0.324083969394795,
  1: -0.38508228041631654,
  2: -0.6769220003059587,
  3: 0.6116762888408679,
  4: 1.0}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [26]:
df_markdown = df[:5].to_markdown()

'|    |     col1 |      col2 |   col3 |      col4 |\n|---:|---------:|----------:|-------:|----------:|\n|  0 | 0.311711 |  0.324084 |      1 |  0.324084 |\n|  1 | 0.520068 | -0.385082 |     -1 | -0.385082 |\n|  2 | 0.54671  | -0.676922 |     -1 | -0.676922 |\n|  3 | 0.184854 |  0.611676 |      1 |  0.611676 |\n|  4 | 0.969585 |  1.031    |      1 |  1        |'

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [27]:
df_markdown


'|    |     col1 |      col2 |   col3 |      col4 |\n|---:|---------:|----------:|-------:|----------:|\n|  0 | 0.311711 |  0.324084 |      1 |  0.324084 |\n|  1 | 0.520068 | -0.385082 |     -1 | -0.385082 |\n|  2 | 0.54671  | -0.676922 |     -1 | -0.676922 |\n|  3 | 0.184854 |  0.611676 |      1 |  0.611676 |\n|  4 | 0.969585 |  1.031    |      1 |  1        |'

Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

|    |     col1 |      col2 |   col3 |      col4 |
|---:|---------:|----------:|-------:|----------:|
|  0 | 0.311711 |  0.324084 |      1 |  0.324084 |
|  1 | 0.520068 | -0.385082 |     -1 | -0.385082 |
|  2 | 0.54671  | -0.676922 |     -1 | -0.676922 |
|  3 | 0.184854 |  0.611676 |      1 |  0.611676 |
|  4 | 0.969585 |  1.031    |      1 |  1        |
